graylog2-server/src/main/java/org/graylog2/rest/resources/streams/destinations/filters/StreamDestinationFiltersResource.java

```diff
@@ -0,0 +1,187 @@
+                paginationParams.getPerPage(),
+                paginationParams.getPage(),
+                dtoId -> isPermitted(RestPermissions.STREAM_DESTINATION_FILTERS_READ, dtoId)
+        );
+
+        return PaginatedResponse.create("elements", paginatedList, paginationParams.getQuery());
+    }
+
+    @GET
+    @Path("/target/{targetId}/filters")
+    @ApiOperation("Get available filter rules for stream and target")
+    public PaginatedResponse<StreamDestinationFilterRuleDTO> getPaginatedFiltersForStreamAndTarget(
+            @ApiParam(name = "streamId", required = true) @PathParam("streamId") @NotBlank String streamId,
+            @ApiParam(name = "targetId", required = true) @PathParam("targetId") @NotBlank String targetId,
+            @ApiParam(name = "pagination parameters") @BeanParam PaginationParameters paginationParams
+    ) {
+        checkPermission(RestPermissions.STREAMS_EDIT, streamId);
+        checkStream(streamId);
+
+        final var paginatedList = filterService.findPaginatedForStreamAndTarget(
+                streamId,
+                targetId,
+                paginationParams.getQuery(),
+                Sorts.ascending(StreamDestinationFilterRuleDTO.FIELD_TITLE),
+                paginationParams.getPerPage(),
+                paginationParams.getPage(),
+                dtoId -> isPermitted(RestPermissions.STREAM_DESTINATION_FILTERS_READ, dtoId)
+        );
+
+        return PaginatedResponse.create("elements", paginatedList, paginationParams.getQuery());
+    }
+
+    @GET
+    @Path("/filters/{filterId}")
+    @ApiOperation("Get filter rule for given ID")
+    public Response getFilter(@ApiParam(name = "streamId", required = true) @PathParam("streamId") @NotBlank String streamId,
+                              @ApiParam(name = "filterId", required = true) @PathParam("filterId") @NotBlank String filterId) {
+        checkPermission(RestPermissions.STREAMS_EDIT, streamId);
+        checkPermission(RestPermissions.STREAM_DESTINATION_FILTERS_READ, filterId);
+        checkStream(streamId);
+
+        final var dto = filterService.findById(filterId)
+                .orElseThrow(() -> new NotFoundException("Filter not found"));
+
+        return Response.ok(wrapDto(dto)).build();
+    }
+
+    @POST
+    @Path("/filters")
+    @ApiOperation("Create new filter rule")
+    @AuditEvent(type = AuditEventTypes.STREAM_DESTINATION_FILTER_CREATE)
+    public Response createFilter(@ApiParam(name = "streamId", required = true) @PathParam("streamId") @NotBlank String streamId,
+                                 @ApiParam(name = "JSON Body", required = true) @Valid StreamDestinationFilterRuleDTO dto) {
+        checkPermission(RestPermissions.STREAMS_EDIT, streamId);
+        checkPermission(RestPermissions.STREAM_DESTINATION_FILTERS_CREATE);
+        checkStream(streamId);
+
+        return Response.ok(wrapDto(filterService.create(streamId, dto))).build();
```

Comment:
I think we should validate also here and on put

Reply:
Aye, good catch!

Reply:
Done in https://github.com/Graylog2/graylog2-server/pull/19739/commits/dd9e7d8b2d70451c28cfc16e8f6871b4404aaffb.